libvast/vast/flatbuffer.hpp

```diff
@@ -19,17 +19,38 @@
 
 namespace vast {
 
+/// Determines whether the FlatBuffers table is a root or a child table, i.e.,
+/// whether the only data contained in the owned chunk is the table itself
+/// (root) or other the table is just part of a bigger root table (child).
+enum class flatbuffer_type {
+  root,  ///< The table type is a root type, or a nested FlatBuffers table.
+  child, ///< The table is sliced from a root table.
+};
+
+/// Determines whether the FlatBuffers table has buffer- or pointer-like
+/// behavior, i.e., whether comparison operators compare the buffer or the table
+/// pointer.
+enum class flatbuffer_semantics {
+  buffer,  ///< Comparison operators compare the buffer contents bytewise. This
+           ///< is only supported for root tables.
+  pointer, ///< Comparison operators compare the table pointers.
+};
```

Comment:
It still feels to me like this complicates the interface for not too much gain (the caller could just call `memcmp()` on any chunk for a buffer-semantics comparison), but if you want to keep it in I won't object further.

It's also a bit weird that by inheriting the `semantics` from the parent, many flatbuffers will end up with the combination of `semantics == buffer && type == child`, which isn't really supported. Maybe it would even make sense to just make a `template<flatbuffer_semantics> bool compare()` function, so that the user has to explicitly select at the call site which semantics he wants.

Reply:
Do we have any table other than `type` that we put into an unordered container as a key? For which tables does the buffer behavior make more sense than the pointer behavior?

Personally I prefer having the buffer behavior everywhere, but I can see where you're coming from. We could also introduce some form of `flatbuffer_base`, and then have a `flatbuffer` and a `flatbuffer_ptr` inherit from that and add the functionality specific to them.

Reply:
I'm not sure I understand the connection to unordered containers?

Reply:
Take `type` for example. We often use it as a key in unordered containers, so they need to have a well-defined `operator<`. If two types have identical buffer contents, we would want them to compare equal, and that also needs to be reflected in the `operator<`.

I think exposing pointer semantics for the comparison operators is rarely what we want.

Reply:
We've decided to straight up remove the comparison operators because we couldn't agree on their semantics.